### ISSUE TITLE
Prerelease not correct with channel: false

### ISSUE THREAD

**Issue by mbrehme** [semantic-release/semantic-release#1499]
### Config:
```
{
	"branches": [
		"master",
		{
			"name": "dev",
			"prerelease": true,
			"channel": false
		},
		{
			"name": "qa",
			"prerelease": true,
			"channel": false
		},
		{
			"name": "beta",
			"prerelease": true,
			"channel": false
		}
	],

	"tagFormat": "release-${version}",
	
	"plugins": [
		"@semantic-release/commit-analyzer",
		"@semantic-release/release-notes-generator",
		"@semantic-release/github"
	]
}
```

### Current-Last-Releases-Tags
release-8.0.0-dev.60
release-8.0.0-beta.1

## Expected behavior

I want to add another release from beta branch and expect it to be release-8.0.0-beta.2.
Instead: it tries to add release-8.0.0-dev.61

I found out that the git notes on dev.60 and beta.1 is  `{"channels":[null]}` which seems correct.

## Environment

- **semantic-release** version: 17.0.4
- CI environment: Node v12.16.1
- Plugins used: 
- **semantic-release** configuration: see current behaviour
- CI logs: `Found git tag release-8.0.0-dev.60 associated with version 8.0.0-dev.60 on branch beta`

**Comment by mbrehme** [comment]
If I try without channel I have the problem that on merge requests the given commit is tagged just for one of the prerelease branches e.g.:

#comit1337 --> Tags: release-1.33.7-dev.69, release-1.33.7-beta.2  <-- But only has the Bote for {channels: ["dev"]}

This also leads to the error message that the tag release-1.33.7-beta.2 already exist because the last found tag is: release-1.33.7-beta.1

**Comment by mbrehme** [comment]
@gr2m how can we solve this issue?

**Comment by gr2m** [comment]
Did you rebase your `beta` branch? If so, check out https://semantic-release.gitbook.io/semantic-release/support/troubleshooting#release-not-found-release-branch-after-git-push-force

**Comment by mbrehme** [comment]
No we don't rebase in our project. It was a normal merge in github.

From my point of view the problem is following:

**Without-Channel**
1) Trrigger a release on Dev -> Tag has the note added `{channels: ["dev"]}`
2) Merge the Dev -> Beta
3) Triggers a release on Beta -> Tag has the note **overwritten** `{channels: ["beta"]}`
4) New change on Dev triggers new release -> it finds the last dev tag with channel in which the last is ignored cause of the note overwrite
5) It tries to recreate the tag which was overwritten therefore the error tag already exist

**With-Channel**
1) Trigger a release on Dev -> Tag has the note added `{channels: [null]}`
2) Merge the Dev -> Beta
3) Triggers a release on Beta -> Search for the last Tag with channel === null which is the dev release
4) It tries to create a "Dev-Tag"

Thank you for your help @gr2m  !!!
P.S.: The troubleshooting helps to restore it for once but the error reoccurs directly on the next merge

**Comment by gr2m** [comment]
I'll have to wait for @pvdlg who implemented the channels feature, I don't have the time right now to dig into it, I'm sorry. Your analysis is sure helpful!

**Comment by pvdlg** [comment]
I'll take a look, but probably not before this weekend.

**Comment by pvdlg** [comment]
@mbrehme just to be sure I understand, when setting `channel` to `false` you want the channel name to determined based on the branch name right? Not to have all your pre-releases to be published on the `@latest` dis-tag?

**Comment by mbrehme** [comment]
This would be one solution.
Another would be when no channel ist set, it appends channels in the notes instead of overwriting. So the release detect the latest of the right channel.

I think the overwriting is an bug and has no further feature requiring this. Therfore I would say it would be better to fix this. 

Thanks a lot for your your time and energy!

**Comment by pvdlg** [comment]
Not sure I'm following...My question was about what you are trying to do.
You want your pre-releases to be published on dist-tags `@beta`, `@qa`, `@dev` or all of them on `@latest` ?

**Comment by mbrehme** [comment]
For me the tags are important (release-8.0.0-dev.46 or release-8.0.0-beta.2) the dist-tags are not important in my case. Just to have the right tags and releases on git, divided by prerelease branch

**Comment by pvdlg** [comment]
I'm sorry I'm really confused and can't figure out why you configure `channel` to `false`.

[Per the documentation](https://github.com/semantic-release/semantic-release/blob/master/docs/usage/workflow-configuration.md#channel) setting the channel to `false` is for publishing to the default channel (i.e `@latest`).

Why don't you configure them as `qa`, `dev` and `beta`?

**Comment by mbrehme** [comment]
Sorry für confusion:

**My Goal:**
- Have Release two tags based on Prerelease Branch eg.: release-8.0.0-dev.46 and release-8.0.0-beta.2 for the same commit after an merge of dev > beta

**Approach-A:**
0) no specific channel settings in config
1) Push Change to Dev 
2) Release is triggered: release-8.0.0-dev.46 (git note channels: dev) 
3) Merge dev to beta 
4) Release is triggered: release-8.0.0-beta.2 ( git note is overwritten with channels: beta because it is the same commit I think)
5) Push new change to dev
6) Release is triggered: release-8.0.0-dev.46 (because the last tag found for this channel was release-8.0.0-dev.45 because the right one was overwritten => **Error Tag already exist**

To overcome this issue I thought I can use 

**Approach-B:**
0) Set channel to false in config for each prerelease branch
1) Push Change to dev
2) Release is triggered: release-8.0.0-dev.46 (git note channels: null) 
3) Merge dev to beta
4) Release is triggered: release-8.0.0-dev.47 (because the last with channel === 0 is release-8.0.0-dev.46) => **release tag dev.47 instead of release.2**

**Comment by pvdlg** [comment]
> Have Release two tags based on Prerelease Branch eg.: release-8.0.0-dev.46 and release-8.0.0-beta.2 for the same commit after an merge of dev > beta

That's not how semantic-release works. We create 1 tag per version released and each version correspond to one particular codebase.
`8.0.0-dev.46` and `8.0.0-beta.2` are 2 different releases with a different version number. They can't correspond to the same commit.
You can't re-release the same thing twice. That's why we support release channels. So the same version can be made available to different group of users based on the dist-tag.

Basically the way it works is:
- When a release is made:
  - a tag is created
  - the package is published on the dist-tag corresponding to that branch
- When you merge the commit associated with the release tag:
  - no new release is done (because you can't release the same thing twice)
  - no new tag is created
  - the existing release is made available on the dis-tag corresponding to the target branch

**Comment by mbrehme** [comment]
For us our continious delivery is based on git tags.

In my Szenario we test the Dev release and if everything works we merge to beta.

How would you archive my goal with semantic release?

Both approaches does not work for us. Does this lead to an feature request or is it possible to "fix" the note overwriting to have eg the note for the commit to have channels: dev, beta.

Again thanks for your help

**Comment by pvdlg** [comment]
Form [Workflow configuration doc](https://github.com/semantic-release/semantic-release/blob/master/docs/usage/workflow-configuration.md#pre-release-branches):
> When merging commits associated with an existing release, semantic-release will treat them as pushed commits and publish a new release if necessary, but it will never add those releases to the distribution channel corresponding to the pre-release branch.

That's because the role of pre-release branch is to allow making several breaking changes commit during the development phase of a big, multi commit feature, without increasing the version number. 

What you are describing (notes with both channel after a merge) is the behavior or regular release branches (ie master, next etc..)

The scope of semantic-release is to make releases on channels but not to determine on environment to deploy them.
It seems that what you want to do is to deploy and test a release on a dev environment, then deploy it and test on a beta environment, then deploy it and test on a qa environment. Is that right?
semantic-release only manage the releases part but doesn't decide where to deploy them.

You could use regular release branches, but they are limited to 3 maximum and are meant to allow you to maintain different branches of the codebase at the same time, not to manage the deployment of version.

Unfortunately, I don't think managing deployments will ever be in the scope of semantic-release. So I would recommend to manage that outside of semantic-release.
Nothing prevent you to create additional tags in your repo, as long as the naming doesn't conflict with the version tags.

For example when you are ready to deploy a version to QA, instead of triggering the deployment based on a merge, just add a tag `deploy-x.y.x-qa` to that release and have you deployment system to listen to that to deploy it

**Comment by mbrehme** [comment]
Okay understand. I thought this is what Prerelses are for cause the merging is also described here: https://github.com/semantic-release/semantic-release/blob/master/docs/recipes/pre-releases.md#publishing-pre-releases

And it almost works perfectly, the only thing required for making it to work correctly would be to fix/change the git note behavior to add the channel instead of overwriting.

**Comment by pvdlg** [comment]
> Okay understand. I thought this is what Prerelses are for cause the merging is also described here: https://github.com/semantic-release/semantic-release/blob/master/docs/recipes/pre-releases.md#publishing-pre-releases

This doc (like all semantic-release docs) is about publishing releases, not deploying them on various environment. That is out of semantic-release scope.

> And it almost works perfectly, the only thing required for making it to work correctly would be to fix/change the git note behavior to add the channel instead of overwriting.

Pre-releases works as they are meant to: allowing to make several breaking changes commit during the development phase of a big, multi commit feature, without increasing the version number.

What you describe as "works perfectly" is your specific use case, deployment management, which is not in the scope of semantic-release. Making the change you are requesting would prevent pre-release branches to work for what they are meant to.

**Comment by mbrehme** [comment]
Okay, thanks for making clear.

I think the system would be more robust if the error of already existing tag would be prevented.

1) Push dev
2) Release dev (automatic through ci)
3) Merge dev -> beta
4) Release beta (automatic through ci)
5) Push dev
6) Release throws error cause it tries to recreate the tag created through the beta release

With the addition of modify the addNote method to read the note and append the channel to the tag/commit and altering the code for picking the last release based on channel AND prerelease branch, would make the whole system more robust IMHO.

Thanks for your time and help!!! Am I right that there will be no further fixing changing?


Mit freundlichen Grüßen
Martin Brehme
Am 27. März 2020, 06:29 +0100 schrieb Pierre Vanduynslager <notifications@github.com>:

**Comment by mbrehme** [comment]
JFYI: I ended up writing an patch which makes sure the last release is based on corresponding prerelease tag.

For anyone interested this is the patch:
```
diff --git a/node_modules/semantic-release/lib/get-last-release.js b/node_modules/semantic-release/lib/get-last-release.js
index f8cc5ed..7ac6ef7 100644
--- a/node_modules/semantic-release/lib/get-last-release.js
+++ b/node_modules/semantic-release/lib/get-last-release.js
@@ -1,6 +1,6 @@
 const {isUndefined} = require('lodash');
 const semver = require('semver');
-const {makeTag, isSameChannel} = require('./utils');
+const {makeTag, isSameChannel, isSamePrerelease} = require('./utils');
 
 /**
  * Last release.
@@ -30,7 +30,7 @@ module.exports = ({branch, options: {tagFormat}}, {before} = {}) => {
   const [{version, gitTag, channels} = {}] = branch.tags
     .filter(
       tag =>
-        ((branch.type === 'prerelease' && tag.channels.some(channel => isSameChannel(branch.channel, channel))) ||
+        ((branch.type === 'prerelease' && tag.channels.some(channel => isSameChannel(branch.channel, channel)) && isSamePrerelease(tag.gitTag, branch.prerelease)) ||
           !semver.prerelease(tag.version)) &&
         (isUndefined(before) || semver.lt(tag.version, before))
     )
diff --git a/node_modules/semantic-release/lib/utils.js b/node_modules/semantic-release/lib/utils.js
index 14d9d02..70f5bbd 100644
--- a/node_modules/semantic-release/lib/utils.js
+++ b/node_modules/semantic-release/lib/utils.js
@@ -77,6 +77,10 @@ function isSameChannel(channel, otherChannel) {
   return channel === otherChannel || (!channel && !otherChannel);
 }
 
+function isSamePrerelease(tag, prerelease) {
+  return tag.indexOf(prerelease) >= 0;
+}
+
 module.exports = {
   extractErrors,
   hideSensitiveValues,
@@ -93,4 +97,5 @@ module.exports = {
   getRange,
   makeTag,
   isSameChannel,
+  isSamePrerelease,
 };
```
